File: python/classes/product_inventory.py
"""Product Inventory — entities with identity, movements as a ledger, current quantity as aggregate."""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional


@dataclass
class Product:
    """A product entity with identity (sku) and mutable fields."""
    sku: str
    name: str
    price: float
    category: str = ""
    reorder_point: int = 0


@dataclass
class Movement:
    """A stock movement — receive or sell."""
    sku: str
    kind: str  # "receive" | "sell"
    quantity: int
    at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    note: str = ""

    @property
    def signed(self) -> int:
        return self.quantity if self.kind == "receive" else -self.quantity


class InsufficientStock(Exception):
    """Raised when a sell exceeds on-hand quantity."""


class UnknownSku(KeyError):
    """Raised when an sku is not registered."""


class Inventory:
    """An inventory of products. Current quantity is aggregated from movements."""

    def __init__(self) -> None:
        self.products: dict[str, Product] = {}
        self.movements: list[Movement] = []

    # --- registration ---
    def add_product(self, product: Product) -> None:
        if product.sku in self.products:
            raise ValueError(f"sku already registered: {product.sku}")
        self.products[product.sku] = product

    def get(self, sku: str) -> Product:
        if sku not in self.products:
            raise UnknownSku(sku)
        return self.products[sku]

    # --- movements ---
    def receive(self, sku: str, quantity: int, note: str = "") -> Movement:
        self.get(sku)
        if quantity <= 0:
            raise ValueError("receive quantity must be positive")
        m = Movement(sku=sku, kind="receive", quantity=quantity, note=note)
        self.movements.append(m)
        return m

    def sell(self, sku: str, quantity: int, note: str = "") -> Movement:
        self.get(sku)
        if quantity <= 0:
            raise ValueError("sell quantity must be positive")
        if self.quantity(sku) < quantity:
            raise InsufficientStock(
                f"{sku}: on hand {self.quantity(sku)}, requested {quantity}"
            )
        m = Movement(sku=sku, kind="sell", quantity=quantity, note=note)
        self.movements.append(m)
        return m

    # --- aggregates ---
    def quantity(self, sku: str) -> int:
        self.get(sku)
        return sum(m.signed for m in self.movements if m.sku == sku)

    def history(self, sku: str) -> list[Movement]:
        self.get(sku)
        return [m for m in self.movements if m.sku == sku]

    def total_value(self) -> float:
        return sum(p.price * self.quantity(p.sku) for p in self.products.values())

    def restock_needed(self) -> list[Product]:
        """Products whose current quantity is at or below their reorder point."""
        return [p for p in self.products.values()
                if p.reorder_point > 0 and self.quantity(p.sku) <= p.reorder_point]

    def by_category(self, category: str) -> list[Product]:
        return [p for p in self.products.values() if p.category == category]

    def summary(self) -> dict:
        return {
            "products": len(self.products),
            "movements": len(self.movements),
            "total_value": round(self.total_value(), 2),
            "restock_count": len(self.restock_needed()),
        }


# --- demo ---
if __name__ == "__main__":
    inv = Inventory()
    inv.add_product(Product("SKU-001", "Widget", 9.99, "tools", reorder_point=5))
    inv.add_product(Product("SKU-002", "Gadget", 24.50, "tools", reorder_point=3))
    inv.add_product(Product("SKU-003", "Sprocket", 2.25, "parts", reorder_point=20))

    inv.receive("SKU-001", 50, note="initial stock")
    inv.receive("SKU-002", 10, note="initial stock")
    inv.receive("SKU-003", 15, note="initial stock")
    inv.sell("SKU-001", 3)
    inv.sell("SKU-002", 8)

    print(inv.summary())
    for p in inv.restock_needed():
        print(f"  restock: {p.sku} {p.name} (on hand {inv.quantity(p.sku)}, reorder at {p.reorder_point})")

File: go/classes/product_inventory.go
// Product Inventory — entities with identity, movements as a ledger, current quantity as aggregate.
package classes

import (
	"errors"
	"fmt"
	"time"
)

// Product is an entity with identity (SKU) and mutable fields.
type Product struct {
	SKU           string
	Name          string
	Price         float64
	Category      string
	ReorderPoint  int
}

// MovementKind is "receive" or "sell".
type MovementKind string

const (
	MovementReceive MovementKind = "receive"
	MovementSell    MovementKind = "sell"
)

// Movement is a stock event recorded to the ledger.
type Movement struct {
	SKU      string
	Kind     MovementKind
	Quantity int
	At       time.Time
	Note     string
}

// Signed returns the movement's effect on quantity (+receive, -sell).
func (m Movement) Signed() int {
	if m.Kind == MovementReceive {
		return m.Quantity
	}
	return -m.Quantity
}

// Inventory errors.
var (
	ErrUnknownSKU         = errors.New("unknown sku")
	ErrDuplicateSKU       = errors.New("duplicate sku")
	ErrNonPositiveQty     = errors.New("quantity must be positive")
	ErrInsufficientStock  = errors.New("insufficient stock")
)

// Inventory aggregates movements to report current quantity per product.
type Inventory struct {
	Products  map[string]Product
	Movements []Movement
}

// NewInventory builds an empty inventory.
func NewInventory() *Inventory {
	return &Inventory{Products: make(map[string]Product)}
}

// AddProduct registers a product.
func (i *Inventory) AddProduct(p Product) error {
	if _, ok := i.Products[p.SKU]; ok {
		return fmt.Errorf("%w: %s", ErrDuplicateSKU, p.SKU)
	}
	i.Products[p.SKU] = p
	return nil
}

// Get returns the registered product or an error.
func (i *Inventory) Get(sku string) (Product, error) {
	p, ok := i.Products[sku]
	if !ok {
		return Product{}, fmt.Errorf("%w: %s", ErrUnknownSKU, sku)
	}
	return p, nil
}

// Receive adds stock.
func (i *Inventory) Receive(sku string, quantity int, note string) error {
	if _, err := i.Get(sku); err != nil {
		return err
	}
	if quantity <= 0 {
		return ErrNonPositiveQty
	}
	i.Movements = append(i.Movements, Movement{
		SKU: sku, Kind: MovementReceive, Quantity: quantity, At: time.Now(), Note: note,
	})
	return nil
}

// Sell removes stock, failing if insufficient.
func (i *Inventory) Sell(sku string, quantity int, note string) error {
	if _, err := i.Get(sku); err != nil {
		return err
	}
	if quantity <= 0 {
		return ErrNonPositiveQty
	}
	onHand, _ := i.Quantity(sku)
	if onHand < quantity {
		return fmt.Errorf("%w: %s on hand %d, requested %d", ErrInsufficientStock, sku, onHand, quantity)
	}
	i.Movements = append(i.Movements, Movement{
		SKU: sku, Kind: MovementSell, Quantity: quantity, At: time.Now(), Note: note,
	})
	return nil
}

// Quantity is the aggregate of all movements for a sku.
func (i *Inventory) Quantity(sku string) (int, error) {
	if _, err := i.Get(sku); err != nil {
		return 0, err
	}
	total := 0
	for _, m := range i.Movements {
		if m.SKU == sku {
			total += m.Signed()
		}
	}
	return total, nil
}

// History returns the movement list for a sku.
func (i *Inventory) History(sku string) ([]Movement, error) {
	if _, err := i.Get(sku); err != nil {
		return nil, err
	}
	var out []Movement
	for _, m := range i.Movements {
		if m.SKU == sku {
			out = append(out, m)
		}
	}
	return out, nil
}

// TotalValue sums price × quantity across all products.
func (i *Inventory) TotalValue() float64 {
	total := 0.0
	for _, p := range i.Products {
		q, _ := i.Quantity(p.SKU)
		total += p.Price * float64(q)
	}
	return total
}

// RestockNeeded returns products at or below their reorder point.
func (i *Inventory) RestockNeeded() []Product {
	var out []Product
	for _, p := range i.Products {
		if p.ReorderPoint <= 0 {
			continue
		}
		q, _ := i.Quantity(p.SKU)
		if q <= p.ReorderPoint {
			out = append(out, p)
		}
	}
	return out
}

// ByCategory filters products.
func (i *Inventory) ByCategory(category string) []Product {
	var out []Product
	for _, p := range i.Products {
		if p.Category == category {
			out = append(out, p)
		}
	}
	return out
}
